Shake feature

firmfreez opened the pull request:
Add shake feature to app

firmfreez committed:
Add retainedState support

firmfreez committed:
Removed .idea files

firmfreez committed:
Add shake feature

jumpstreet13 commented on app/src/main/java/com/firmfreez/android/ballofdesires/view/main/MainActivity.kt:
```diff
@@ -25,9 +28,38 @@ class MainActivity : AppCompatActivity(), MainView {
         binding = ActivityMainBinding.inflate(layoutInflater)
         setContentView(binding.root)
 
+        binding.progress.max = 100
```
лучше такие цифры выносить в константы, чтобы можно было сразу увидеть смысл, который в них заложен

firmfreez replied:
Ок, поправлю.

jumpstreet13 commented on app/src/main/java/com/firmfreez/android/ballofdesires/view/main/MainActivity.kt:
```diff
@@ -25,9 +28,38 @@ class MainActivity : AppCompatActivity(), MainView {
+        shaker = ShakeListener(this)
+        shaker?.shakeListener = object : ShakeListener.OnShakeListener {
+            override fun onShake(progress: Float) {
+                binding.progress.progress = (progress * 100).roundToInt()
```
лучше такие цифры выносить в константы, чтобы можно было сразу увидеть смысл, который в них заложен

firmfreez replied:
Ок, поправлю.

jumpstreet13 commented on app/src/main/java/com/firmfreez/android/ballofdesires/view/main/MainActivity.kt:
```diff
@@ -25,9 +28,38 @@ class MainActivity : AppCompatActivity(), MainView {
+            override fun onComplete() {
+                getPresenter().onStartBtnClicked()
+                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
+                    vibrator?.vibrate(VibrationEffect.createOneShot(1000,0))
```
лучше такие цифры выносить в константы, чтобы можно было сразу увидеть смысл, который в них заложен

firmfreez replied:
Ок, поправлю, а константы лучше выносить как private val или в companion object { const val } ??

jumpstreet13 replied:
так как это константы времени компиляции, то лучше сделать const val, он тогда ее в байткоде сделает статической

firmfreez replied:
А в чем заключается приемущества статики в байткоде? Только если мы будем использовать эту переменную где либо еще кроме этого класса, он будет подтягивать то же значение, а не созлавать новое?

jumpstreet13 replied:
Ну, из плюсов могу перечислить то, что она будет загружаться в память вместе с классом один раз, а не каждый раз с каждым инстансом 
плюс статика потокобезопасна, так как грузится во время того как загружается класс 

оба эти плюса в разрезе активити не кажутся такими уж значительными, но это можно даже назвать более идиоматичный подходом

jumpstreet13 commented on app/src/main/java/com/firmfreez/android/ballofdesires/view/main/MainPresenter.kt:
```diff
@@ -22,6 +23,12 @@ class MainPresenter: BasePresenter<MainView>() {
+
+        //Восстанавливаем данные на форме
+        retainedState?.get<YesNoModel>(OLD_DATA)?.let {
+            Log.d(TAG, retainedState?.get<YesNoModel>(OLD_DATA).toString())
```
можешь чекнуть либу для логирования Timber мб будет интересно и она тебе полезные плюшки даст

firmfreez replied:
Изучил и добавил в проект. Одна из полезных плюшек - это то, что мы "садим" дерево логов при старте приложения. Соотвественно, поставив там условие на BuildConfig.DEBUG, у нас не будет лога на релизных версиях приложения. А какие еще приемущества даёт Timber?

jumpstreet13 replied:
Из самых значительных плюшек могу назвать ,что там можно написать дерево, которое отправляет логи в крашлитику)

firmfreez replied:
О, это прикольно, надо будет как нибудь затестить (может уже на проекте :))

jumpstreet13 commented on app/src/main/java/com/firmfreez/android/ballofdesires/view/main/MainPresenter.kt:
```diff
@@ -32,17 +39,18 @@ class MainPresenter: BasePresenter<MainView>() {
                     view?.setBallAnswer(it)
+                    retainedState?.set(OLD_DATA, it)
                 }, {
                     Log.e(TAG, it.localizedMessage.orEmpty())
```
хорошо бы на будущие пул реквесты продумать свою схему обработки ошибок и как ты их будешь пользователю показывать 
например на случай если интернет отвалился

firmfreez replied:
Готово. Реализовал с помощью диалогов. Теперь где угодно (где есть Router) можно вызвать кастомный диалог. И даже поставить Listener на реакцию пользователя APPLY, CANCEL или DISMISS.

jumpstreet13 commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -32,6 +32,16 @@
+            android:id="@+id/progress"
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="16dp"
```
не супер критичный комментарий, но вот такие стандартные отступы по хорошему в dimens выносить, потом захочешь отступы от края экрана поменять по всей приле, поменяешь их в дименсах и будет счастье

firmfreez replied:
Готово, поменял по всему приложению.

jumpstreet13 commented on app/src/main/java/com/firmfreez/android/ballofdesires/view/main/ShakeListener.kt:
```diff
@@ -0,0 +1,99 @@
+            throw UnsupportedOperationException("Sensor is not supported")
+        }
+        val supported = sensorManager?.registerListener(this, sensorManager?.getDefaultSensor(Sensor.TYPE_ACCELEROMETER), SensorManager.SENSOR_DELAY_GAME)
+//        if(supported == false) {
```
закоменченный код лучше не пушить на пул реквесте

на крайняк если думаешь, что понадобится можешь его застешить в студии там через vcs-git-stash 

а потом когда надо будет сделать unstash

firmfreez replied:
Готово. Просто удалил, они в целом не очень нужны.

jumpstreet13 commented on app/src/main/java/com/firmfreez/android/ballofdesires/view/main/MainActivity.kt:
```diff
@@ -25,9 +28,38 @@ class MainActivity : AppCompatActivity(), MainView {
+        binding.progress.max = 100
+
+        vibrator = (getSystemService(VIBRATOR_SERVICE) as? Vibrator)
+        shaker = ShakeListener(this)
```
всегда когда есть возможность передать кому-то апликейшн контест лучше передать его 

активити контекст может утечь от будущих неаккуртаных изменений в ShakeListenerи будет утечка памяти, а так ты передашь апликейшн и ничего такого не будет 

плюс можно сюда через DI инжектить ShakeListener так будет правильнее раз уж в проекте есть DI

firmfreez replied:
Готово. Заинжектил через DI.

firmfreez committed:
Add constants to All digits in project
Include Timber in project instead utils.Log

firmfreez committed:
Add cicerone to project
Add TextDialogFragment for showing dialogs everywhere on project
Add feature, that showing error dialog when network connection disabled

firmfreez committed:
Remove comments from project
Make border dimensions constants in project

firmfreez committed:
Move shakeListener to DI

jumpstreet13 commented on app/src/main/res/layout/fragment_text_dialog.xml:
```diff
@@ -0,0 +1,87 @@
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="16dp"
+            android:layout_marginTop="16dp"
```
тут константы забыл)

firmfreez replied:
Принял, заменил.

jumpstreet13 commented on app/src/main/java/com/firmfreez/android/ballofdesires/view/dialogs/TextDialogFragment.kt:
```diff
@@ -0,0 +1,126 @@
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        binding = FragmentTextDialogBinding.inflate(inflater, container, false)
```
мне такое в общем норм, я бы не стал коммент оставлять, все читаемо и понятно, но на проектах бывают душные ревьюверы, которые попросили бы декомпозировать этот метод на несколько мелких
так что лучше я тебя подготовлю и к такому) 
декомпозируй плиз на более мелкие методы, чтобы можно было четче логику проследить

firmfreez replied:
Ок, разбил)

firmfreez committed:
Add pull request bug fixes
Redesign TextDialogFragment for best quality display

jumpstreet13 commented on app/src/main/res/layout/fragment_text_dialog.xml:
```diff
@@ -12,18 +12,18 @@
         <TextView
             android:id="@+id/title"
-            android:layout_width="0dp"
+            android:layout_width="200dp"
```
не совсе понял зачем фиксированная ширина у текствьюхи, сегодня вечером обсудим)

firmfreez replied:
Объяснил на митапе)
